### ISSUE TITLE
Question about tslm basic model

### ISSUE THREAD

**Issue by danielhanbitlee** [davidthaler/Walmart_competition_code#2]
Hi David,

First, thank you for posting your code and for the explanations you provided [here](https://www.kaggle.com/c/walmart-recruiting-store-sales-forecasting/discussion/8125) and [here](https://www.kaggle.com/c/walmart-recruiting-store-sales-forecasting/discussion/8033) about your winning methods. I've been learning a lot from them.

This isn't much of an "issue" as it is a conceptual question about the tslm basic model. Here is the code from [here](https://github.com/davidthaler/Walmart_competition_code/blob/master/grouped.forecast.R) that I'm trying to better understand:

```
tslm.basic <- function(train, test){
  horizon <- nrow(test)
  train[is.na(train)] <- 0
  for(j in 2:ncol(train)){
    s <- ts(train[, j], frequency=52)
    model <- tslm(s ~ trend + season)
    fc <- forecast(model, h=horizon)
    test[, j] <- as.numeric(fc$mean)
  }
  test
}`
```
Here is how I am understanding your explanations from the Kaggle posts and the code you provided:

For every department and store combination, fit a multiple linear regression model where the model is as follows:

<a href="https://www.codecogs.com/eqnedit.php?latex=y_t&space;=&space;\beta_0&space;&plus;&space;\beta_1t&space;&plus;&space;\beta_2d_{2,t}&space;&plus;&space;...&space;&plus;&space;\beta_{52}d_{52,t}&space;&plus;&space;\epsilon_t" target="_blank"><img src="https://latex.codecogs.com/gif.latex?y_t&space;=&space;\beta_0&space;&plus;&space;\beta_1t&space;&plus;&space;\beta_2d_{2,t}&space;&plus;&space;...&space;&plus;&space;\beta_{52}d_{52,t}&space;&plus;&space;\epsilon_t" title="y_t = \beta_0 + \beta_1t + \beta_2d_{2,t} + ... + \beta_{52}d_{52,t} + \epsilon_t" /></a>

where <a href="https://www.codecogs.com/eqnedit.php?latex=d_{i,t}&space;=&space;1" target="_blank"><img src="https://latex.codecogs.com/gif.latex?d_{i,t}&space;=&space;1" title="d_{i,t} = 1" /></a> if **t** is in the specified week of year and 0 otherwise. **i** represents the week of year (i = 2 to 52).

If date is in 2010, **t** = week of year
If date is in 2011, **t** = week of year + 52
If date is in 2012, **t** = week of year + 52 * 2
If date is in 2013, **t** = week of year + 52 * 3

Here is an example:

week_of_year = 5, date: 2010-02-05, t = 5
week_of_year = 5, date: 2011-02-04, t = 5 + 52 = 57
week_of_year = 5, date: 2012-02-03, t = 5 + 52 * 2 = 109
week_of_year = 5, date: 2013-02-01, t = 5 + 52 * 3 = 161

Use the train set to fit the linear regression models. Then, for the test set, make predictions on weekly sales using these models. If the dept-store pair in the test set is not found in the train set, fill in the weekly sales with zero.

I ask this because I'm trying to implement this in python and have been having some problems.

Any help would be greatly appreciated.

Best,
Daniel

**Comment by davidthaler** [comment]
Daniel,

That all looks right to me. With that model, I forecast each department-store sales series separately. And the `tslm(s ~ trend + season)` model is as you described it. It has a dummy variable for each week and a linear trend.    

As an aside, that is really a lot of parameters and I don't think I would use this model on data this short. For each week, we have either 2 or 3 years of training data, so each dummy variable has just 2 or 3 points bearing on it.    

Finally, Rob Hyndman's online textbook [Forecasting Principles and Practice](https://otexts.org/fpp2/) is a really nice introduction to practical forecasting. I used it a lot in the Walmart competition. The tslm model is covered in chapter 5 there.

Good luck on your project,

Dave

**Comment by danielhanbitlee** [comment]
Ah got it. Thank you for the helpful response Dave.

Best,
Daniel
